pkg/controller/buildrun/buildrun_controller.go

```diff
@@ -180,12 +181,24 @@ func (r *ReconcileBuildRun) Reconcile(request reconcile.Request) (reconcile.Resu
 				buildRun.Status.Reason = lastTaskRun.Status.Conditions[0].Reason
 			}
 		}
-		buildRun.Status.LatestTaskRunRef = &lastTaskRun.Name
-		buildRun.Status.StartTime = lastTaskRun.Status.StartTime
-		buildRun.Status.CompletionTime = lastTaskRun.Status.CompletionTime
 		if buildRun.Status.BuildSpec == nil {
 			buildRun.Status.BuildSpec = &build.Spec
 		}
+
+		if lastTaskRun.Status.CompletionTime != nil && buildRun.Status.CompletionTime == nil {
+			buildRun.Status.CompletionTime = lastTaskRun.Status.CompletionTime
+
+			// Increase BuildRun count in metrics
+			buildmetrics.BuildRunCountInc(build.Spec.StrategyRef.Name)
+			// Add BuildRun establish and completion times in metrics
+			buildRunCompletionDuring := buildRun.Status.CompletionTime.Time.Sub(buildRun.CreationTimestamp.Time)
+			buildRunEstablishDuring := buildRun.Status.StartTime.Time.Sub(buildRun.CreationTimestamp.Time)
+			buildmetrics.BuildRunEstablishObserve(build.Spec.StrategyRef.Name, buildRun.Namespace, buildRunEstablishDuring)
```

Comment:
You capture this metric on completion. You can capture this earlier by using condition `lastTaskRun.Status.StartTime != nil && buildRun.Status.StartTime == nil`.

Reply:
Yes, I also think about that.
If we use `lastTaskRun.Status.StartTime != nil && buildRun.Status.StartTime == nil`
We can cover more, but still cannot cover all :(
- We cover the buildrun count of taskrun is executed successfully
- We cover the buildrun count of taskrun is executed fail
- But we still cannot cover the buildrun count when fail to create sa, add ownereference or any failure before generate taskrun correctly.

So how should we name the metric? `build_buildruns_taskrun_succeedandfail` ?
So that is why I also suggest to improve our logic with metric in future together.